enhancements/agent-installer/automated-workflow.md

```diff
@@ -0,0 +1,491 @@
+   or [the `oc mirror`
+   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
+3. The cluster admin obtains the OpenShift installer in one of [the
+   usual
+   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
+4. The cluster admin triggers their orchestration tool to run the
+   installer.
+5. The orchestration tool prepares an `install-config.yaml` and
+   separate manifest files with `NMState` content using input data and
+   templates.
+6. The orchestration tool runs `openshift-install create image`.
+7. The installer extracts a copy of the RHCOS ISO from the OpenShift
+   release payload.
+8. The installer combines the inputs provided by the user and produces
+   an Ignition config.
+9. The installer combines the Ignition config and the RHCOS ISO to
+   create the bootable image configured to run the assisted service or
+   agent based on the host that boots the image and writes the new
+   image and cluster credentials to disk.
+10. The orchestration tool copies the ISO to a location that the hosts
+    or VMs can boot from it. For bare metal, that will generally be a
+    web server visible to the management controllers in the bare metal
+    hosts (because the management controllers are unlikely to be on
+    the same network as the external interfaces for the cluster
+    hosts). For vSphere or other hypervisors, that may be a filesystem
+    on the hypervisor host.
+11. The orchestration tool configures the host to boot the ISO. For
+    bare metal, that means configuring the baseboard management
+    controller (BMC) in each host to expose the ISO through the
+    virtual media interface. For vSphere or other hypervisors, similar
+    settings of the VM need to be adjusted.
+12. The orchestration tool boots the host.
+13. The orchestration tool uses `openshift-install wait-for
+    install-complete` to watch the deployment progress and wait for it
+    to complete.
+14. On each host, the startup script in the image selects the correct
+    network settings based on the MAC addresses visible and applies
+    the network settings for each interface.
+15. On node0, the startup script in the image recognizes that it is on
+    node0 and launches the assisted service and the create-cluster
+    service to drive the deployment.
+16. On node0, the create-cluster service waits for the assisted
+    service to start, then creates an InfraEnv using the API.
+17. On node0, the create-cluster service waits for all of the known
+    agents to be registered with the service.
+18. On all nodes (including node0), the startup script in the image
+    connects to the assisted service on node0 to find the InfraEnv ID
+    for the cluster being deployed (there will only be 1). After it
```

Comment:
We're doing it this way now, but is that how we want to keep doing it? Could we pre-generate the ID and embed it in the image?

Reply:
Pre-generating sounds reasonable.

Reply:
The ID is randomly generated within assisted-service: https://github.com/openshift/assisted-service/blob/afd8849feee8c450ba02085f16e723429ae32a6f/internal/bminventory/inventory.go#L3985

To accept a pre-generated id, the /v2/infra-envs REST-API endpoint would need to be changed to allow the id to become one of the parameters. Is that possible?

Reply:
Allowing users to specify the ID would be a security issue.  We use the InfraEnv ID as part of the agent authentication.  Deleting an InfraEnv is our way of revoking the agents' access.  Allowing (potentially malicious) users to create an InfraEnv with the same ID as a deleted one is dangerous.
@carbonin please keep me honest here

Reply:
Yes, the InfraEnv ID is used in our authentication methods.

The agent also requires the InfraEnv ID (and auth token) as input and does not have permissions to list InfraEnvs thus cannot discover what InfraEnvs may be available in the environment.
I did notice this in the non-goals though:
> Details of securing the communication between agents and services to
  ensure only desired hosts join the cluster are left to a future
  enhancement.

What does that mean for your plans to support authentication/https between the agent and service? I see several ways that auth between the agent and service could complicate this procedure.

Reply:
Given that we only intend to support a single InfraEnv in the ephemeral installer, we could pre-generate the ID and pass it in as e.g. an environment variable. It wouldn't have to become part of the user-accessible API.